### ISSUE TITLE
Make externally snapped plugin updates more stable

### ISSUE THREAD

**Issue by bmw** [certbot/certbot#7863]
With [the way externally snapped plugins work](https://github.com/certbot/certbot/blob/master/tools/snap/README.md), Certbot and its dependencies mask the version of Certbot installed in the plugin snap. (See https://github.com/certbot/certbot/issues/7946 about removing the 2nd copy of Certbot altogether, but even if that's resolved, the purpose of this issue remains.) As a result, if the plugin snap is upgraded before the Certbot snap and that plugin requires a new version of Certbot, the plugin would break until Certbot is also updated.

When this problem happens, I expect it to be resolved quickly (unless the user did something like manually held Certbot's updates back), but since as far as I can tell there is no well defined ordering for updates, I expect problems like this to occur fairly regularly for users of externally snapped plugins. I personally don't like this at all and unless we can convenience ourselves that it is a very minor problem, I think we should either fix it or not offer externally snapped plugins. It's just doesn't seem stable enough to me for it to be the official way to install an external plugin with Certbot.

Luckily, we may have a solution here. Robie looked into https://forum.snapcraft.io/t/supported-snap-hooks/3795/ which describes a way to define hooks which if they fail, snaps will either not be updated or they will be rolled back. This needs to be tested and designed a little carefully to avoid something like a badly written plugin snap causing the Certbot snap to never upgrade, but it seems like this could fairly easily solve our problems.

This issue is mainly a research/design issue trying to design a system to avoid these problems with our own or third party plugins, although assuming we find a design we like, we'll need to make sure we implement it once we have Certbot/plugin snaps in this repo.

As of writing this, the instructions at https://github.com/certbot/certbot/blob/master/tools/snap/README.md#use-testing-and-development can be used to create proof-of-concept snap plugins.

**Comment by bmw** [comment]
@ohemorange, I've update the proof of concept instructions so things should work for you now.

If you hit any issues, please let me know!

**Comment by ohemorange** [comment]
tl;dr: Sticking a hook named `post-refresh` into the plugin’s `meta/hooks/` directory that fails if certbot is the wrong version for the newly updated plugin's needs should mostly work. There’s a few potential issues we might run into while implementing this, but most of those have backup solutions detailed below. The major design question is if we’re ok with having this solution not work for backwards-incompatible Certbot releases, and instead being sure to update plugins a release ahead of deprecations in those cases.

This is all also written up [in a Google doc](https://docs.google.com/document/d/1P4mgEMzvsh2n3ipCGgcISFmIvBdisP-Aa9mXtgWIEEc/edit?usp=sharing), in case it's easier to have discussions there.

---

Depending on when exactly metadata is updated in the refresh process, we’ll want either pre-refresh or post-refresh hooks. Either way, failure will roll back the entire refresh, leaving the plugin at its old version until the next refresh attempt (usually about six hours later).

These hooks are added to the plugins. They will check the certbot version, and if it’s not correct, fail (in some way), which will cause the refresh to not happen. Since snap checks for updates 4 times a day, later that day certbot should have been installed and the plugin will continue updating.

If it’s a pre-refresh hook, the hook goes in a file called `meta/hooks/pre-refresh` for the particular plugin.

The contents of the file could look something like this:
```
#!/bin/sh -e

# get certbot version
# get required certbot version for plugin 

If [ certbot does not fit required version ]; then
    echo "Certbot needs to be updated before this plugin can be updated; will try again on next refresh."
    exit 1
fi
```
Then the refresh will not happen.

Again, depending what metadata is updated when the refresh hook is called, we may not be able to access the required certbot version until after the refresh happens. But again, that’s ok, because if we put the above test in a post-refresh hook instead, the refresh will be rolled back entirely:

This hook is executed for the newly installed snap, before starting new services (if applicable). This hook is a good place for any extra actions that need to be performed for the new revision of the snap. It’s also a good place to test whether a refresh has succeeded, because if the test causes the hook to fail, the refresh will be rolled-back and the original state of the snap restored.

The script, whether post- or pre-refresh, will need to access the versions of Certbot compatible with the new version of the plugin. We keep those in `$plugindir/setup.py:install_requires`. This file is probably accessible by the hook script, because the hooks are run from within the context of the snap, but if not, another option to look into could be defining it as metadata somewhere within the snapcraft file, and accessing that. But it’s probably not necessary to do so.

Another potential problem is that the test might for some reason access the version of certbot internal to the plugin. The easiest and best fix for this is to remove that extra copy, since we were planning to do that anyway, rather than trying to figure out how to access the primary version of certbot.

Once that version is removed, we’ll want to make sure that the plugin snap can access the certbot version. This might just work given the existing interface connection hookups, but if not, we’ll probably need to duplicate the slots/plugs from the plugin’s and certbot’s `snapcraft.yml`s under the appropriate headings, like so:
```
hook:
    post-refresh:
        slots: [...]
```
We’re already running within the snap’s environment though, so theoretically we should be able to access it in at least some way, given that plugins can call into certbot’s libraries within their code -- I don’t think we’ll need to define extra interfaces, unless it’s easier to do so than to get the version out of certbot that way.

Assuming that Certbot has not done a backwards-incompatible change, this should theoretically be a complete solution, since the plugin will continue working, just using the last version, for the next six hours.

But given the recent 20.04 nginx update issue, we may want to figure out what to do about non-backwards-compatible breaking changes. Assuming this is an update that we (the certbot team) have put out, theoretically the plugin will only be broken for six hours. The simplest solution is to be ok with that, given how often and ahead of schedule we attempt to renew certificates. If we do that, we absolutely must print out some manner of error message that goes to a place a user would think to check. If we want to print to /var/log/letsencrypt, we’ll have to add a plug to the plugin’s snapcraft.yml, like how certbot connects to the yml. But honestly, if we do go this route, we’re probably fine just echoing the error message, although we should confirm that the error message goes somewhere readable and find out where that is.

More importantly, if we do this, we will want to be careful to stop using deprecated certbot functionality at least one release ahead for plugins. In general, we should theoretically be able to do this, given how far ahead of time we have Certbot start warning about future deprecations. This is what we expect third-party plugin authors to do. I can foresee some potential issues, like emergency updates or forgetting to do this. For the former, we could always do an emergency release process-change and just publish main certbot a bit ahead of the plugins. For the latter, we could add a lint script or something that reminds us that we need to update plugins a release ahead, perhaps gated on having incremented Certbot’s major version number.

I go into all this detail above, because I don’t actually believe it’s possible (and if it is, it’s certainly not possible in a way that’s adherent to the intended usage patterns for snaps) to actually fix the problem of making sure Certbot updates before the plugins update. There’s no interface for refreshing individual plugins; you’re meant to call `snap refresh` and let it do its magic for you. But, this actually IMO gives some context to the larger issue we’re trying to solve here: it may not necessarily be that big of a problem. If all updates always happen batched together, we’re talking an order of minutes where the plugin might be broken. That’s not great, but it’s not necessarily the end of the world. Obviously, I think this should be the fallback option given how not-bad the above steps seem, but it is comforting to keep in mind.

**Comment by bmw** [comment]
The high level plan Erica describes above LGTM and I'm almost certain we can resolve this problem if we want to in the future.

Erica would like to spend a bit more time on this to see if she can get something actually working once we have DNS plugin snaps, so I'm moving this to 1.7 and adding a "blocked" label for now. The issue tracking creating DNS plugin snaps is https://github.com/certbot/certbot/issues/7672.

**Comment by bmw** [comment]
I think this is unblocked, but @ohemorange please correct me if there were other pieces you'd prefer we wait on.

**Comment by ohemorange** [comment]
I agree that this is now unblocked.

**Comment by ohemorange** [comment]
Hey, I  think we can do this without adding any interfaces, by calling `snap info certbot` and checking the `installed` line to get the certbot version.

**Comment by bmw** [comment]
Maybe! Can a classically confined plugin snap run `snap` to inspect other snaps installed on the host?

**Comment by ohemorange** [comment]
I'm pretty sure that's what "Hooks are called with no parameters. When a hook needs to request or modify information in within snapd, they can do so via the snapctl tool, which is always available within a snap’s environment. See Using the snapctl tool for further details." means. From https://snapcraft.io/docs/supported-snap-hooks. Will test though.

**Comment by ohemorange** [comment]
Conversely, it is harder to get the required certbot version than expected, because `setup.py` is not packaged in the snap. I can think of a few ways to get it:
A. Put the required certbot version in the plugin's snap meta/env/whatever manually
B. A, but s/manually/automatically while snapping/
C. Stick setup.py in the plugin snap
D. Run `certbot plugins` and see if that fails.

A. Very easy to fix now, but we'll have to remember to update that version in a *third* place when making future changes (or we could grab and set it during snapping).
B. I think this should be doable. It's another env variable floating around, but that's not that big of a problem.
C. Feels weird to do, but shouldn't actually be too hard to modify the script to copy it in for fun while snapping...
D. I don't love this, because I'm worried that it'll have false positives and actually fail when you try to run it.

I'm leaning towards B.

**Comment by ohemorange** [comment]
C is implemented on `plugin-stability`

**Comment by ohemorange** [comment]
Ok, here's where I am. I think that we'll need both a pre-refresh and post-refresh hook to do this, and they'll communicate via a `snapctl` config setting. That's because `pre-refresh` runs in the old snap, and `post-refresh` runs in the new snap. That part is straightforward. Also, once I staged `setup.py`, that should just be available to grab the required certbot version in the `post-refresh` hook. It's getting the certbot version that's currently giving me trouble. I've set up a plug and a slot and some mechanisms for only checking things on refresh (not first install), but I can't figure out where my syntax is wrong. Currently getting `- Run pre-refresh hook of "certbot-dns-dnsimple" snap if present (run hook "pre-refresh": grep: /snap/certbot-dns-dnsimple/x2/certbot-exec: Is a directory)`. If the answer doesn't come to me in a dream I'll probably ask on the forums.

**Comment by ohemorange** [comment]
What if it were a write interface rather than a read, and certbot wrote its own version out after refreshing

Certbot technically doesn't even need an interface to do that, since it's classically confined, I think...

I also wonder if there's something wrong with the content interface, since why create the `CERTBOT_PLUGINS` env var if the libraries appeared where they were supposed to on certbot's end of the interface?

**Comment by ohemorange** [comment]
Hey wait what if we just set the certbot version as an attribute of the connection https://snapcraft.io/docs/interface-hooks

This might mean that certbot needs to refresh once before the plug-in can ever refresh, but that's fine.

**Comment by basak** [comment]
> If it’s a pre-refresh hook, the hook goes in a file called meta/hooks/pre-refresh for the particular plugin.

May I suggest doing this the other way round? What if it's the certbot snap's pre- or post- refresh hook that is responsible for doing the check? Then, for externally developed and published plugins, you don't have to rely on the third party correctly implementing the hook for the check to work.

Presumably certbot (without snaps) will fail to run if a plugin is installed that is incompatible with it? In that case, a preliminary post-refresh hook for the certbot snap would simply run certbot with some no-op parameter (eg. `--help`?) that nevertheless loads plugins. If it succeeds, then the hook succeeds. If it fails, then the hook fails, causing snapd to revert the update. This would catch other miscellaneous problems as well as the specific plugin incompatibility issue.

One thought on pre-refresh vs. post-refresh is that with post-refresh you have the opportunity to fix bugs and/or behaviour in the hook itself. What happens if a pre-refresh hook has a bug? Is it possible with pre-refresh for a buggy script to require user invention to fix? It isn't clear to me from the documentation exactly which hook version will be used.

**Comment by ohemorange** [comment]
Ah yes, we have now veered quite a bit from the original design you're quoting from @basak. A few things:

- The main case we're worried about here is that the plugin wants to add functionality, but that functionality also needs to be added in certbot. I don't think putting the check in certbot's code will fix that. Basically our assumption is that it's always ok for Certbot to update, but sometimes a plugin might want to hold off on updating.

- You're right about fixing bugs in the pre-refresh hook. The only way I've found to fix that in my testing is to manually uninstall and reinstall the plugin.

- Unfortunately, we'll probably actually have to use both pre and post refresh hooks. Pre-refresh hooks run on the previous version of the snap, and post-refresh hooks run on the new version of the snap (and I'm not convinced they have a set order of which of them runs first, and if they do, I think it might actually be the post-refresh hook, annoyingly). We need the post-refresh hook because that will have the version of certbot that the updated plugin requires, which the previous version of the plugin won't have. But we need the pre-refresh hook as well, because post-refresh hooks don't seem to have access to anything from connections. I would love to be wrong about that, because it would make life much easier. Hopefully I just haven't found the syntax for that yet.

**Comment by ohemorange** [comment]
> Hey wait what if we just set the certbot version as an attribute of the connection https://snapcraft.io/docs/interface-hooks

Well, I don't see a way for this to work. Having tried it, the error message helpfully indicates that "interface attributes can only be read during the execution of interface hooks." This is of course not on the documentation page.

**Comment by ohemorange** [comment]
> I've set up a plug and a slot and some mechanisms for only checking things on refresh (not first install), but I can't figure out where my syntax is wrong. Currently getting - Run pre-refresh hook of "certbot-dns-dnsimple" snap if present (run hook "pre-refresh": grep: /snap/certbot-dns-dnsimple/x2/certbot-exec: Is a directory). If the answer doesn't come to me in a dream I'll probably ask on the forums.

The answer didn't come to me in a dream, but looking at this again after PTO and catching-up-on-everything-else-I-missed-during-PTO, I'm pretty sure I was at no point actually connecting the new interface I'd added. 🤦

Going to check that now, this is why we take PTO, kids!

**Comment by ohemorange** [comment]
I think I found the problem. 😭

> The content sharing interface bind-mounts the slot to the plug's target. To that end, the target parameter must be pointing to an existing directory (bind mounts need to mount on top of something just like any other mount). 

https://askubuntu.com/questions/841004/cannot-get-basic-content-interface-example-working-with-snapcraft

**Comment by ohemorange** [comment]
WELL. FUN FACT! the argument to `read` in the content interface has to be a directory. Who would have guessed?? I have no citation because as far as I can tell, this isn't stated anywhere. Just my own testing.

**Comment by ohemorange** [comment]
GREAT news! The pre-refresh hook, which scared me with its difficulty in fixing bugs, is not necessary after all.

Current status: all the snapcraft stuff works! Yay! Things are connected!

Next steps:
- compare versions properly in bash
- clean up all the debugging code
